csi/controller_server.go

```diff
@@ -381,56 +309,78 @@ func (cs *ControllerServer) ControllerPublishVolume(ctx context.Context, req *cs
+	//  should be handled by the processing of the api return codes
 
-	existVol, err := cs.apiClient.Volume.ById(req.GetVolumeId())
+	volume, err := cs.apiClient.Volume.ById(req.GetVolumeId())
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
-	if existVol == nil {
+	if volume == nil {
 		msg := fmt.Sprintf("ControllerPublishVolume: the volume %s not exists", req.GetVolumeId())
 		logrus.Warn(msg)
 		return nil, status.Error(codes.NotFound, msg)
 	}
 
-	if existVol.RestoreRequired {
-		return nil, status.Errorf(codes.Aborted, "The volume %s is restoring backup", req.GetVolumeId())
+	if volume.Frontend != string(types.VolumeFrontendBlockDev) {
+		return nil, status.Errorf(codes.InvalidArgument, "ControllerPublishVolume: there is no block device frontend for volume %s", req.GetVolumeId())
 	}
 
-	if existVol.State == string(types.VolumeStateAttaching) || existVol.State == string(types.VolumeStateDetaching) {
-		return nil, status.Errorf(codes.Aborted, "The volume %s is %s", req.GetVolumeId(), existVol.State)
+	if requiresSharedAccess(volume, volumeCapability) {
+		volume, err = cs.updateVolumeAccessMode(volume, types.AccessModeReadWriteMany)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	if existVol.Frontend != string(types.VolumeFrontendBlockDev) {
-		return nil, status.Errorf(codes.InvalidArgument, "ControllerPublishVolume: there is no block device frontend for volume %s", req.GetVolumeId())
+	// TODO: JM Restore should be handled by the volume attach call, consider returning `codes.Aborted`
+	// TODO: JM should readiness be handled by the caller?
+	//  Most of the readiness conditions are covered by the attach, except auto attachment which requires changes to the design
+	//  should be handled by the processing of the api return codes
+	if !volume.Ready {
+		return nil, status.Errorf(codes.Aborted, "The volume %s in state %v is not ready for workloads",
+			req.GetVolumeId(), volume.State)
 	}
 
-	isAttached := existVol.State == string(types.VolumeStateAttached)
-	waitForController := isAttached && (len(existVol.Controllers) == 0 || existVol.Controllers[0].Endpoint == "")
-	if !existVol.Ready || waitForController {
-		return nil, status.Errorf(codes.Aborted, "The volume %s in state %v is not ready for workloads, is waiting on controller %v",
-			req.GetVolumeId(), existVol.State, waitForController)
+	// TODO: JM if volume is already attached to a different node, return code `codes.FailedPrecondition`
+	//  this should be handled by the processing of the api return code
+	if !requiresSharedAccess(volume, volumeCapability) && volume.State == string(types.VolumeStateAttached) && volume.Controllers[0].HostId != req.GetNodeId() {
+		return nil, status.Errorf(codes.FailedPrecondition, "The volume %s cannot be attached to the node %s since it is already attached to the node %s",
+			req.GetVolumeId(), req.GetNodeId(), volume.Controllers[0].HostId)
 	}
```

Comment:
Maybe we can let this csi server blindly send attach calls even without checking readiness or frontend. But I am not sure if the upper level will do something special for some special error code like `codes.FailedPrecondition`.

Reply:
Yeah we should but for v1.1.1 I don't have time to implement structured error for the backend.
So I handled this specific case in the csi driver for the moment with the TODO.
FailedPrecondition is dedicated to the multi attach condition.
Technically we could just call the api and return internal server error which is less specific.